File: app/src/main/java/com/akari/ppx/xp/hook/code/fun/InfoHook.java
package com.akari.ppx.xp.hook.code.fun;

import com.akari.ppx.common.utils.InfoUtils;
import com.akari.ppx.common.utils.XSP;
import com.akari.ppx.xp.hook.code.SuperbHook;

import java.util.ArrayList;
import java.util.List;

import de.robv.android.xposed.XC_MethodHook;

import static com.akari.ppx.common.constant.Const.AUTHOR_ID;
import static com.akari.ppx.common.constant.Prefs.DIY_ENABLE;
import static com.akari.ppx.common.constant.Prefs.MODIFY_MESSAGE_COUNTS;
import static com.akari.ppx.common.constant.Prefs.PUNISH_ENABLE;
import static de.robv.android.xposed.XposedHelpers.callMethod;
import static de.robv.android.xposed.XposedHelpers.callStaticMethod;
import static de.robv.android.xposed.XposedHelpers.findClass;
import static de.robv.android.xposed.XposedHelpers.getLongField;
import static de.robv.android.xposed.XposedHelpers.getObjectField;
import static de.robv.android.xposed.XposedHelpers.newInstance;
import static de.robv.android.xposed.XposedHelpers.setLongField;
import static de.robv.android.xposed.XposedHelpers.setObjectField;

public class InfoHook extends SuperbHook {
	@Override
	protected void onHook(ClassLoader cl) {
		final boolean diyEnable = XSP.get(DIY_ENABLE), punishEnable = XSP.get(PUNISH_ENABLE);
		final InfoUtils mInfo = new InfoUtils();
		final String[] mName = new String[]{""};
		hookMethod("com.sup.android.uikit.avatar.FrameAvatarView", "setUserInfo", "com.sup.android.mi.usercenter.model.UserInfo", int.class, new XC_MethodHook() {
			private void setField(Object obj, String fieldName, String value) {
				if (!"".equals(value)) {
					try {
						setLongField(obj, fieldName, Long.parseLong(value));
					} catch (Exception ignored) {
					}
				}
			}

			@Override
			protected void beforeHookedMethod(MethodHookParam param) {
				Object userInfo = param.args[0];
				if (getLongField(userInfo, "id") == AUTHOR_ID) {
					final String desc = "皮皮虾助手开发者";
					Object certifyInfo = newInstance(findClass("com.sup.android.mi.usercenter.model.UserInfo$CertifyInfo", cl));
					callMethod(certifyInfo, "setCertifyType", 2);
					callMethod(certifyInfo, "setDescription", desc);
					setObjectField(userInfo, "certifyInfo", certifyInfo);
					setObjectField(userInfo, "description", desc);
				}
				if (diyEnable) {
					if ("".equals(mName[0])) {
						try {
							mName[0] = (String) getObjectField(callMethod(callStaticMethod(findClass("com.sup.android.module.usercenter.UserCenterService", cl), "getInstance"), "getMyUserInfo"), "name");
						} catch (Exception ignored) {
							mName[0] = "";
						}
					}
					if (getObjectField(userInfo, "name").equals(mName[0])) {
						if (punishEnable) {
							Object Punishment = newInstance(findClass("com.sup.android.mi.usercenter.model.UserInfo$Punishment", cl));
							setObjectField(Punishment, "shortDesc", "已被移送小黑屋");
							ArrayList list = new ArrayList();
							list.add(Punishment);
							setObjectField(userInfo, "punishmentList", list);
						}
						int certifyType = Integer.parseInt(mInfo.getCertifyType());
						if (certifyType > 0 && certifyType <= 4) {
							Object certifyInfo = newInstance(findClass("com.sup.android.mi.usercenter.model.UserInfo$CertifyInfo", cl));
							callMethod(certifyInfo, "setCertifyType", certifyType);
							String desc = mInfo.getCertifyDesc();
							callMethod(certifyInfo, "setDescription", !"".equals(desc) ? desc : "阿勒，忘填描述了？");
							setObjectField(userInfo, "certifyInfo", certifyInfo);
						}
						String name = mInfo.getUserName();
						if (!"".equals(name))
							setObjectField(userInfo, "name", name);
						String desc = mInfo.getDescription();
						if (!"".equals(desc))
							setObjectField(userInfo, "description", desc);
						setField(userInfo, "likeCount", mInfo.getLikeCount());
						setField(userInfo, "followersCount", mInfo.getFollowersCount());
						setField(userInfo, "followingCount", mInfo.getFollowingCount());
						setField(userInfo, "point", mInfo.getPoint());
					}
				}
			}
		});
		hookMethod(MODIFY_MESSAGE_COUNTS, "com.google.gson.Gson", "fromJson", String.class, Class.class, new XC_MethodHook() {
			@Override
			protected void afterHookedMethod(MethodHookParam param) {
				if (diyEnable && "com.sup.android.m_message.data.t".equals(((Class) param.args[1]).getName())) {
					for (Object i : (List) getObjectField(getObjectField(param.getResult(), "data"), "count_map")) {
						setObjectField(i, "count", 100);
					}
				}
			}
		});
	}
}

File: app/src/main/java/com/akari/ppx/xp/hook/misc/InfoHook.kt
@file:Suppress("unused")

package com.akari.ppx.xp.hook.misc

import com.akari.ppx.data.Const.APP_NAME
import com.akari.ppx.data.Const.AUTHOR_ID
import com.akari.ppx.data.XPrefs
import com.akari.ppx.utils.*
import com.akari.ppx.xp.Init.cl
import com.akari.ppx.xp.hook.BaseHook

class InfoHook : BaseHook {
    override fun onHook() {
        val (isCustomize, isModifyMsgCounts, isEnterBlackHouse) = listOf<Boolean>(
            XPrefs("customize"),
            XPrefs("modify_message_counts"),
            XPrefs("enter_black_house")
        )
        var (certifyType, certifyDesc, username, description, likeCount, followersCount, followingCount, point) = listOf<String>(
            XPrefs("certify_type"),
            XPrefs("certify_desc"),
            XPrefs("username"),
            XPrefs("description"),
            XPrefs("like_count"),
            XPrefs("followers_count"),
            XPrefs("following_count"),
            XPrefs("point")
        )
        val avatarViewClass = "com.sup.android.uikit.avatar.FrameAvatarView".findClass(cl)
        val method = avatarViewClass.declaredMethods.find { m ->
            m.parameterTypes.size == 2 && m.parameterTypes[0].name == "com.sup.android.mi.usercenter.model.UserInfo"
                    && m.parameterTypes[1] == Int::class.java
        }?.name
        avatarViewClass.hookBeforeMethod(
            method,
            "com.sup.android.mi.usercenter.model.UserInfo",
            Int::class.java
        ) { param ->
            val userInfo = param.args[0]
            val certifyInfoClass =
                "com.sup.android.mi.usercenter.model.UserInfo\$CertifyInfo".findClass(cl)
            if (userInfo.getLongField("id") == AUTHOR_ID) {
                certifyInfoClass.new().apply {
                    callMethod("setCertifyType", 2)
                    callMethod("setDescription", "${APP_NAME}开发者")
                }.let {
                    userInfo.setObjectField("certifyInfo", it)
                }
            }
            isCustomize.check(true) {
                "com.sup.android.module.usercenter.UserCenterService".findClass(cl)
                    .callStaticMethod("getInstance")
                    ?.callMethod("getMyUserInfo")?.getObjectFieldAs<String>("name")
                    .check(userInfo.getObjectFieldAs("name")) {
                        if (isEnterBlackHouse) {
                            userInfo.setObjectField("punishmentList", arrayListOf<Any>().apply {
                                "com.sup.android.mi.usercenter.model.UserInfo\$Punishment".findClass(
                                    cl
                                ).new().apply {
                                    setObjectField("shortDesc", "已被移送小黑屋")
                                }.let(::add)
                            })
                        }
                        certifyType.checkUnless("") {
                            toInt().checkIf({ this in 1..3 }) {
                                certifyInfoClass.new().apply {
                                    callMethod("setCertifyType", this@checkIf)
                                    if (certifyDesc.isEmpty())
                                        certifyDesc = "阿勒，忘填描述了？"
                                    callMethod("setDescription", certifyDesc)
                                }.let { userInfo.setObjectField("certifyInfo", it) }
                            }
                        }
                        username.checkIf({ isNotEmpty() }) {
                            userInfo.setObjectField("name", this)
                        }
                        description.checkIf({ isNotEmpty() }) {
                            userInfo.setObjectField("description", this)
                        }
                        with(userInfo) {
                            runCatching { likeCount.toLong() }.getOrNull()
                                ?.let { setLongField("likeCount", it) }
                            runCatching { followersCount.toLong() }.getOrNull()
                                ?.let { setLongField("followersCount", it) }
                            runCatching { followingCount.toLong() }.getOrNull()
                                ?.let { setLongField("followingCount", it) }
                            runCatching { point.toLong() }.getOrNull()
                                ?.let { setLongField("point", it) }
                        }
                    }
            }
        }
        "com.google.gson.Gson".findClass(cl)
            .hookAfterMethod("fromJson", String::class.java, Class::class.java) { param ->
                if (isModifyMsgCounts && (param.args[1] as Class<*>).name == "com.sup.android.m_message.data.t") {
                    param.result.getObjectField("data")?.getObjectFieldAs<List<*>>("count_map")
                        ?.forEach {
                            it.setObjectField("count", 100)
                        }
                }
            }
    }
}